api/src/main/java/org/apache/iceberg/expressions/Predicate.java

```diff
@@ -19,15 +19,44 @@
 
 package org.apache.iceberg.expressions;
 
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 public abstract class Predicate<T, R extends Reference> implements Expression {
   private final Operation op;
   private final R ref;
   private final Literal<T> literal;
+  private final Set<Literal<T>> literalSet;
 
   Predicate(Operation op, R ref, Literal<T> lit) {
     this.op = op;
     this.ref = ref;
     this.literal = lit;
+    if (lit == null || lit == Literals.aboveMax() || lit == Literals.belowMin() ||
```

Comment:
`aboveMax` and `belowMin` are only created by `Literal#to` during the binding process. I think it is better to handle them in binding instead of here in the constructor.

Reply:
👍